File: .claude/skills/decomp-func/scripts/census.py
#!/usr/bin/env -S uv run python3
"""レベル2・3の未一致関数をサイズ昇順で列挙する。

リポジトリのルートから実行すること。出力形式は TSV: size<TAB>name<TAB>inc_path
- レベル2: NAKED + INCFUNC (C シグネチャはあるが実装なし)
- レベル3: NON_MATCH + INCFUNC (C 実装はあるが一致しない)
"""

import re
import subprocess
import sys
from pathlib import Path

REPO = Path(
    subprocess.run(["git", "rev-parse", "--show-toplevel"], capture_output=True, text=True).stdout.strip()
)


def collect_targets():
    """src/ の各 .c ファイルを行ごとに走査して NAKED / NON_MATCH 関数を収集する。

    NAKED の場合:    同じ行に INCFUNC がある
    NON_MATCH の場合: #else ブロック内に INCFUNC がある
    """
    targets = {}  # name -> inc の相対パス

    for src_file in sorted((REPO / "src").glob("*.c")):
        state = None        # "NON_MATCH" | None
        current_name = None
        in_else = False     # #else に入ったかどうか

        for line in src_file.read_text(encoding="utf-8", errors="replace").splitlines():
            s = line.strip()

            # 先頭が NAKED または NON_MATCH の関数定義行を検出する
            m = re.match(r"(NAKED|NON_MATCH)\s+\S+\s+(\w+)\s*\(", s)
            if m:
                level = m.group(1)
                name = m.group(2)
                if level == "NAKED":
                    # NAKED は同じ行に INCFUNC がある
                    inc_m = re.search(r'INCFUNC\("(asm/func/[^"]+\.inc)"\)', s)
                    if inc_m:
                        targets[name] = inc_m.group(1)
                else:
                    # NON_MATCH: #else ブロック内の INCFUNC を待つ
                    state = "NON_MATCH"
                    current_name = name
                    in_else = False
                continue

            if state == "NON_MATCH" and current_name:
                if s == "#else":
                    in_else = True
                elif s == "#endif":
                    # ブロック終端
                    state = None
                    current_name = None
                    in_else = False
                elif in_else:
                    inc_m = re.search(r'INCFUNC\("(asm/func/[^"]+\.inc)"\)', s)
                    if inc_m:
                        targets[current_name] = inc_m.group(1)
                        state = None
                        current_name = None
                        in_else = False

    return targets


def estimate_size(inc_path):
    """inc ファイルの命令行からバイトサイズを推定する。"""
    est = 0
    for line in inc_path.read_text(encoding="utf-8", errors="replace").splitlines():
        s = line.strip()
        if (
            not s
            or s.startswith(("@", ".align", ".include", ".syntax", ".text", ".thumb",
                              "thumb_func", "arm_func"))
            or s.endswith(":")
        ):
            continue
        if s.startswith(".4byte"):
            est += 4
        elif s.startswith((".short", ".2byte")):
            est += 2
        elif s.startswith(".byte"):
            est += 1
        elif not s.startswith("."):
            est += 2
    return est


def main():
    targets = collect_targets()
    if not targets:
        print("対象関数が見つかりませんでした", file=sys.stderr)
        return 1

    rows = []
    for name, inc_rel in targets.items():
        inc_path = REPO / inc_rel
        size = estimate_size(inc_path) if inc_path.exists() else 0
        rows.append((size, name, inc_rel))

    rows.sort()
    for size, name, inc in rows:
        print(f"{size}\t{name}\t{inc}")
    print(f"# {len(rows)} functions remaining", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .claude/skills/decomp-func/scripts/census.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { Command } from "@cliffy/command";
import * as path from "@std/path";

// e.g. .claude/skills/decomp-func/scripts/census.ts src/enemy_080ef84c.c
// e.g. .claude/skills/decomp-func/scripts/census.ts src/enemy_manager.c src/enemy_080ef84c.c
// e.g. .claude/skills/decomp-func/scripts/census.ts src/**/*.c
const main = () => {
  new Command()
    .name("census.ts")
    .description("未一致関数 (NAKED / NON_MATCH) の関数をサイズ昇順で列挙する。出力は TSV: size<TAB>level<TAB>name<TAB>inc_path")
    .argument("<files...:string>", "探索対象の .c のパス (glob で複数渡せる)")
    .action((_opts, ...targets: string[]) => {
      census(getRepoRoot(), targets);
    })
    .parse(Deno.args);
};

const die = (msg: string): never => {
  console.error(msg);
  Deno.exit(1);
};

const resolveCFile = (target: string): string => {
  const abs = path.isAbsolute(target) ? target : path.resolve(target);
  let stat: Deno.FileInfo;
  try {
    stat = Deno.statSync(abs);
  } catch {
    die(`見つかりません: ${target}`);
  }
  if (!stat!.isFile) die(`.c のパスを渡してください (ディレクトリは不可): ${target}`);
  if (path.extname(abs) !== ".c") die(`.c ではありません: ${target}`);
  return abs;
};

const incRe = /INCFUNC\("(asm\/func\/[^"]+\.inc)"\)/;
// 型は複数トークンになりうる (`NAKED EntityD854Node* Foo(`) ので、`(` の直前の識別子を関数名とする
const defRe = /^(NAKED|NON_MATCH)\s+[^()]*?(\w+)\s*\(/;

type Level = "NAKED" | "NON_MATCH";

// 関数名 -> レベルと inc の相対パス
const collectTargets = (files: string[]): Map<string, { level: Level; inc: string }> => {
  const targets = new Map<string, { level: Level; inc: string }>();

  for (const file of files) {
    let pending: string | null = null; // #else の INCFUNC を待っている NON_MATCH 関数
    let inElse = false;

    for (const line of Deno.readTextFileSync(file).split("\n")) {
      const s = line.trim();

      const def = s.match(defRe);
      if (def) {
        if (def[1] === "NAKED") {
          // NAKED は同じ行に INCFUNC がある
          const inc = s.match(incRe);
          if (inc) targets.set(def[2], { level: "NAKED", inc: inc[1] });
        } else {
          pending = def[2];
          inElse = false;
        }
        continue;
      }

      if (pending === null) continue;
      if (s === "#else") {
        inElse = true;
      } else if (s === "#endif") {
        pending = null;
        inElse = false;
      } else if (inElse) {
        const inc = s.match(incRe);
        if (inc) {
          targets.set(pending, { level: "NON_MATCH", inc: inc[1] });
          pending = null;
          inElse = false;
        }
      }
    }
  }

  return targets;
};

// inc ファイルの命令行からバイトサイズを推定する
const estimateSize = (incPath: string): number => {
  let est = 0;
  for (const line of Deno.readTextFileSync(incPath).split("\n")) {
    const s = line.trim();
    if (!s || s.endsWith(":")) continue;
    if (/^(@|\.align|\.include|\.syntax|\.text|\.thumb|thumb_func|arm_func)/.test(s)) continue;
    if (s.startsWith(".4byte")) est += 4;
    else if (s.startsWith(".short") || s.startsWith(".2byte")) est += 2;
    else if (s.startsWith(".byte")) est += 1;
    else if (!s.startsWith(".")) est += 2;
  }
  return est;
};

const census = (repo: string, targets: string[]) => {
  const files = [...new Set(targets.map(resolveCFile))].sort();

  const found = collectTargets(files);
  if (found.size === 0) die("対象関数が見つかりませんでした");

  const rows = [...found].map(([name, { level, inc }]) => {
    const incPath = path.join(repo, inc);
    let size = 0;
    try {
      size = estimateSize(incPath);
    } catch {
      size = 0; // inc が無い場合
    }
    return { size, level, name, inc };
  });

  rows.sort((a, b) => a.size - b.size || a.name.localeCompare(b.name));
  for (const r of rows) console.log(`${r.size}\t${r.level}\t${r.name}\t${r.inc}`);
  console.error(`# ${rows.length} functions remaining`);
};

const getRepoRoot = (): string => {
  const { stdout } = new Deno.Command("git", { args: ["rev-parse", "--show-toplevel"] }).outputSync();
  return new TextDecoder().decode(stdout).trim();
};

if (import.meta.main) main();
